include/RAJA/policy/hip/WorkGroup/Vtable.hpp

```diff
@@ -109,6 +109,23 @@ inline const Vtable_T* get_Vtable(hip_work<BLOCK_SIZE, Async> const&)
   return &vtable;
 }
 
+/*!
+* Explicit BLOCKS_PER_SM version.
+* Populate and return a Vtable object where the
+* call operator is a device function
+*/
+template < typename T, typename Vtable_T, size_t BLOCK_SIZE, size_t BLOCKS_PER_SM, bool Async >
```

Comment:
According to AMD docs this should be WARPS_PER_EU instead of BLOCKS_PER_SM.
https://rocmdocs.amd.com/en/latest/Programming_Guides/Kernel_language.html?highlight=launch_bounds#launch-bounds

Reply:
That's correct, I've applied the conversion suggested by ROCM where `WARPS_PER_EU = (THREADS_PER_BLOCK * BLOCKS_PER_SM) / 32` at the sites of the HIP `__launch_bounds__`. This way, we can keep `BLOCKS_PER_SM` as the standard.

Reply:
I understand that that will keeps things the same across cuda and hip, but I'm not sure that's our goal. I thought we were trying to expose the underlying programming model and this is a place where hip and cuda differ. @rhornung67 @trws what do you think?

Reply:
I agree that we should be faithful to the underlying PM since users should be aware of basic concepts like CUDA warp size is 32 threads, HIP "wavefront" size is 64 threads, etc.

Reply:
So that leaves us to either specify WARPS_PER_EU directly or do something in between like BLOCKS_PER_CU. I see where the documentation mentions that formula, but it also says that how many EU there are per CU is not known at compile time so it's not possible to convert BLOCKS_PER_CU to WARPS_PER_EU at compile time without assuming the value of EU_PER_CU of a certain architecture.

Reply:
We should be sure to add a NOTE to the documentation about this so people are aware.

Reply:
Perhaps we should just expose it directly to the user. It's technically `MIN_WARPS_PER_EU`, so we can just do what's recommended by the documentation and default it to 1 such that it's never a problem, and then warn the user if they try something else.

Reply:
What do we mean by 'EU' here? It should be CU, right? AMD CU (compute unit) and NVIDIA SM (streaming multiprocessor) are analogous; i.e., the smallest functional unit on a GPU. What am I missing?

Reply:
I'm fine with `MIN_WARPS_PER_EU`. EU is execution unit (SIMD), see the link in the first post.
I'm not sure that formula makes much sense for the GPUs we're using anyway as we have 4 EUs per CU. 
If we have 256 THREADS_PER_BLOCK and 1 BLOCKS_PER_SM that formula yields 8 WARPS_PER_EU when it should be just 1.

Reply:
Annoying addition of EU to mean execution unit aside, I'd say we should pass the lockstep width through.  If we pretend it's 32, it's entirely possible for a user to write code with it that will deadlock that would be correct if we used the correct size.

Reply:
Just to be sure we're on the same page, as far as I understand It isn't SIMD, it's more like hyperthreads or similar.  It's the functional unit within a CU which has the resources we're trying to reason about, and there may be 1 or 4 of them per CU.  This is not a terribly easy thing to reason about, since it's another level in the hierarchy that we aren't used to.  SIMD is *below* the EU.  I think the Nvidia side technically has this too, but they don't expose the resources of the EUs separately from the resources of a CU.  In general they try relatively hard to hide the fact that each SM actually schedules 4 quarter-warps simultaneously on sub-elements from the user.  Either way, apparently we're stuck with this.

Reply:
That is true. I'm not sure what passing the lockstep size through would look like. I'm pretty sure we have cuda warp and hip wavefront sizes defined somewhere.